cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp

```diff
@@ -0,0 +1,977 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#ifndef __INTERNAL_TSNE_GRADIENT_DESCENT_FPT_CPP__
+#define __INTERNAL_TSNE_GRADIENT_DESCENT_FPT_CPP__
+
+#include "algorithms/tsne/tsne_gradient_descent.h"
+#include "data_management/data/numeric_table.h"
+#include "data_management/data/csr_numeric_table.h"
+#include "services/daal_defines.h"
+#include "services/env_detect.h"
+#include "src/externals/service_math.h"
+#include "src/externals/service_dispatch.h"
+#include "src/services/service_data_utils.h"
+#include "src/services/service_utils.h"
+#include "src/services/service_defines.h"
+#include "src/data_management/service_numeric_table.h"
+#include "src/algorithms/service_error_handling.h"
+
+using namespace daal::data_management;
+using namespace daal::internal;
+
+namespace daal
+{
+namespace algorithms
+{
+namespace internal
+{
+template <typename DataType, CpuType cpu>
+class TlsMax : public daal::TlsMem<DataType, cpu, services::internal::ScalableCalloc<DataType, cpu> >
+{
+public:
+    typedef daal::TlsMem<DataType, cpu, services::internal::ScalableCalloc<DataType, cpu> > super;
+    TlsMax(size_t n) : super(n) {}
+    void reduceTo(DataType * res, size_t n)
+    {
+        bool bFirst = true;
+        this->reduce([=, &bFirst](DataType * ptr) -> void {
+            if (!ptr) return;
+            if (bFirst)
+            {
+                for (size_t i = 0; i < n; ++i) res[i] = ptr[i];
+                bFirst = false;
+            }
+            else
+            {
+                for (size_t i = 0; i < n; ++i) res[i] = services::internal::max<cpu, DataType>(res[i], ptr[i]);
+            }
+        });
+    }
+};
+
+template <typename IdxType, daal::CpuType cpu>
+services::Status maxRowElementsImpl(const size_t * row, const IdxType N, IdxType & nElements)
+{
+    TlsMax<IdxType, cpu> maxTlsData(1);
+    const IdxType nThreads    = threader_get_threads_number();
+    const IdxType sizeOfBlock = services::internal::min<cpu, IdxType>(256, N / nThreads + 1);
+    const IdxType nBlocks     = N / sizeOfBlock + !!(N % sizeOfBlock);
+
+    daal::threader_for(nBlocks, nBlocks, [&](IdxType iBlock) {
+        const IdxType iStart = iBlock * sizeOfBlock;
+        const IdxType iEnd   = services::internal::min<cpu, IdxType>(N, iStart + sizeOfBlock);
+        IdxType * localMax   = maxTlsData.local();
+        for (IdxType i = iStart; i < iEnd; ++i)
+        {
+            localMax[0] = services::internal::max<cpu, IdxType>(localMax[0], IdxType((row[i + 1] - row[i])));
+        }
+    });
+    maxTlsData.reduceTo(&nElements, 1);
+
+    return services::Status();
+}
+
+template <typename IdxType, typename DataType, daal::CpuType cpu>
+services::Status boundingBoxKernelImpl(DataType * posx, DataType * posy, const IdxType N, const IdxType nNodes, DataType & radius)
```

Comment:
Naming of variables posX, posY, ... were not aligned. Please fix it in all places